automation/Extension/webext-instrumentation/src/background/http-instrument.ts

```diff
@@ -468,15 +468,49 @@ export class HttpInstrument {
       }
     }
     */
-    update.top_level_url = escapeUrl(tab.url);
+    update.top_level_url = escapeUrl(this.getDocumentUrlForRequest(details));
     update.parent_frame_id = details.parentFrameId;
     update.frame_ancestors = escapeString(
       JSON.stringify(details.frameAncestors),
     );
-
     this.dataReceiver.saveRecord("http_requests", update);
   }
 
+  /**
+   * Code taken and adapted from
+   * https://github.com/EFForg/privacybadger/pull/2198/files
+   *
+   * Gets the URL for a given request's top-level document.
+   *
+   * The request's document may be different from the current top-level document
+   * loaded in tab as requests can come out of order:
+   *
+   * @param {WebRequestOnBeforeSendHeadersEventDetails} details
+   *
+   * @return {?String} the URL for the request's top-level document
+   */
+  private getDocumentUrlForRequest(
+    details: WebRequestOnBeforeSendHeadersEventDetails,
+  ) {
+    let url = "";
+
+    if (details.type === "main_frame") {
+      // Url of the top-level document itself.
+      url = details.url;
+    } else if (details.hasOwnProperty("frameAncestors")) {
+      // In case of nested frames, retrieve url from top-most ancestor.
+      // If frameAncestors == [], request comes from the top-level-document.
+      url = details.frameAncestors.length
+        ? details.frameAncestors[details.frameAncestors.length - 1].url
+        : details.documentUrl;
+    } else {
+      // type != 'main_frame' and frameAncestors == undefined
+      // For example service workers: https://bugzilla.mozilla.org/show_bug.cgi?id=1470537#c13
+      url = details.documentUrl;
```

Comment:
I just found out about another type of requests that falls into this category, which are requests made by regular Worker scripts (not ServiceWorkers).
They match the two conditions `type != 'main_frame' && frameAncestors == undefined`, but  contrary to ServiceWorkers they DO have a tabId, which corresponds to the tab that spawned them.
For those, my patch actually changed how they are attributed. They used to be attributed to the url of the tab, now they are assigned their own url as top-level-url.
Example: 
`'http://localtest.me/test_page.html'` spawns a worker script located at `'http://localtest.me/scripts/worker.js'`
The scripts makes an xmlHttpRequest to fetch some other resource. 
This request used to be recored with `top-level-url = 'http://localtest.me/test_page.html'`, since that's the url of the tab.
Now it is recorded with `top-level-url = 'http://localtest.me/scripts/worker.js'`.
I'm not sure that is desired, if the url of the parent document of the Worker script is available through the tabId.

I stumbled upon this while writing a test for the ServiceWorker case. Unfortunately, ServiceWorkers have to be served over https, so I wasn't able to write such a test. I suppose it's highly non-trivial to make the test infrastructure support https requests, right?

> It's moving us in the right direction, even if it's unclear what URL will be used for ServiceWorkers.

It is actually clear, it's the url of the ServiceWorker script itself. So if  the `worker.js` script in my example were a ServiceWorker instead of a regular Worker, it would still be attributed with the same top-level-url.

Reply:
> They match the two conditions type != 'main_frame' && frameAncestors == undefined, but contrary to ServiceWorkers they DO have a tabId, which corresponds to the tab that spawned them.

That's good to know, but ultimately if we were to retrieve the top-level URL through the Tab ID we should still run the risk of an incorrect attribution.  Since we have the `tabId` and the navigation events associated with that `tabId` via the navigation instrumentation I think a consumer of the data could still reconstruct the tab url from that data, so there's no need to approximate it here.

> I stumbled upon this while writing a test for the ServiceWorker case. Unfortunately, ServiceWorkers have to be served over https, so I wasn't able to write such a test. I suppose it's highly non-trivial to make the test infrastructure support https requests, right?

This would be out of scope for this testing, but it probably isn't too complex. But for this case I think we can side-step it by just accessing the test page using `localhost` instead of `localtest.me`. `localhost` [is a secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts). We prefer using the test domain since it ensures we don't test our instrumentation on a privileged origin, but here it seems necessary to use `localhost`.